kivy/uix/carousel.py

```diff
@@ -361,6 +353,18 @@ def _insert_visible_slides(self, _next_slide=None, _prev_slide=None):
         else:
             self._next = None
 
+        expected_value = self.loop and (len(self.slides) == 2)
+        if self._prev_equals_next != expected_value:
+            from kivy.logger import Logger
+            Logger.error(
+                "Carousel: I expected '_prev_equals_next' to be already "
```

Comment:
This smells to me like it should be an `assert`. Logging errors is generally discouraged because it trains people to ignore them, and if this path is hit, users would see an error without much for them to do. This seems like if this path is hit it would be a bug in kivy if I read this correctly, hence an `assert` is more appropriate.

Reply:
This may be related to the order of evaluation. I mean things like this:

```python
from kivy.lang import Builder
from kivy.factory import Factory
from kivy.properties import StringProperty


class TestEvaluationOrder(Factory.Widget):
    prop1 = StringProperty()
    prop2 = StringProperty()

    def apply_class_lang_rules(self, *args, **kwargs):
        def handler(c, self, *args):
            print(c, ':', self.prop2 == self.prop1 + ' prop2')

        self.fbind('prop1', handler, 'A')
        r = super().apply_class_lang_rules(*args, **kwargs)
        self.fbind('prop1', handler, 'B')
        return r

KV_CODE = '''
<TestEvaluationOrder>:
    prop2: self.prop1 + ' prop2'

TestEvaluationOrder:
'''
w =  Builder.load_string(KV_CODE)
w.prop1 = 'prop1'
```

```
A : False
B : True
```

> Logging errors is generally discouraged because it trains people to ignore them

got it